### PR TITLE
Fixed #26900 -- Fixed crash accessing deferred FileFields.

### PR DESCRIPTION
**Pull request by timgraham**
https://code.djangoproject.com/ticket/26900

### CODE REVIEW

**Comment by samuelmaudo** [comment]
if I understand well the patch, it has a problem with ImageFields. Since image file is requiered whenever the model is instantiated, defer a ImageField results in an extra database query for each object in the QuerySet. That could generate hundreds of database queries.

**Comment by timgraham** [comment]
Do you have time to put together a test that demonstrates the issue?

**Comment by timgraham** [comment]
Here's one that results in `RecursionError` with this patch but works on 1.9:
```python
def test_defer(self):
    self.PersonModel.objects.create(name='Joe', mugshot=self.file1)
    self.PersonModel.objects.create(name='Joe', mugshot=self.file1)
    self.PersonModel.objects.create(name='Joe', mugshot=self.file1)
    qs = self.PersonModel.objects.defer('mugshot')
    for obj in qs:
        self.assertEqual(obj.name, 'Joe')
```
Thanks, I'll investigate further.

**Comment by timgraham** [comment]
Please check if the updated patch solves that issue.

File: tests/model_fields/test_imagefield.py [hunk 1]
```diff
@@ -194,6 +194,12 @@ def test_pickle(self):
         loaded_p = pickle.loads(dump)
         self.assertEqual(p.mugshot, loaded_p.mugshot)
 
+    def test_defer(self):
+        self.PersonModel.objects.create(name='Joe', mugshot=self.file1)
+        qs = list(self.PersonModel.objects.defer('mugshot'))
```
Comment (samuelmaudo) [inline tests/model_fields/test_imagefield.py]: The extra queries could be generated in this line (199), as this is where the instancies are created, so so you should check the number of queries is one.

**Comment by samuelmaudo** [comment]
The patch seems right but the test is wrong since it doesn't check the number of database queries where the intances are created.

**Pull request merged**